### PR TITLE
Giotto learn

### PR DESCRIPTION
**Pull request by giotto-learn**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by giotto-learn** [comment]
Dear Conda-forge team (*in particular those with python/C++ expertise*),
I am struggling to build the wheels of my package on windows, as you can see from my many attempts. 
Can you please give me some hint? Notice that on my repo, <https://github.com/giotto-learn/giotto-learn> the wheels are building for windows as well <https://dev.azure.com/giotto-learn/giotto-learn/_build/results?buildId=14>. I believe the problem is the vs2017 compiler, but I am unsuccessfully trying to make it work. Here the log file for the azure pipeline:
[28.txt](https://github.com/conda-forge/staged-recipes/files/3707083/28.txt)

**Comment by marcelotrevisani** [comment]
> Dear Conda-forge team (_in particular those with python/C++ expertise_),
> I am struggling to build the wheels of my package on windows, as you can see from my many attempts.
> Can you please give me some hint? Notice that on my repo, https://github.com/giotto-learn/giotto-learn the wheels are building for windows as well https://dev.azure.com/giotto-learn/giotto-learn/_build/results?buildId=14. I believe the problem is the vs2017 compiler, but I am unsuccessfully trying to make it work. Here the log file for the azure pipeline:
> [28.txt](https://github.com/conda-forge/staged-recipes/files/3707083/28.txt)

Someone opened another PR for the same package
https://github.com/conda-forge/staged-recipes/pull/9774

File: conda-forge.yml [hunk 1]
```diff
@@ -1,2 +1,2 @@
 max_py_ver: '37'
-max_r_ver: '35'
\ No newline at end of file
+max_r_ver: '35'
```
Comment (marcelotrevisani) [inline conda-forge.yml]: Please, revert this file

File: recipes/giotto-learn/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  version: {{ version }}
+
+source:
+  git_url: https://github.com/{{ name }}/{{ name }}.git
+
+build:
+  number: 0
+  skip: True  # [py2k]
+  script:
+    - git submodule update --init --recursive
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: Why do you need these submodules?
is that a dependency?

File: recipes/giotto-learn/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    - cmake -G "Visual Studio 15 2017"  # [win]
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: if you need it just for ``vs2017``
you should modify skip for different versions of vs

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]:
```suggestion
  skip: true  # [win and vs<14]
```

File: recipes/giotto-learn/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    - "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}  # [not win]
+    - vs2017_win-64  # [win]
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: Please remove this line and also the selector

File: recipes/giotto-learn/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  number: 0
+  skip: True  # [py2k]
+  script:
+    - git submodule update --init --recursive
+    - cmake -G "Visual Studio 15 2017"  # [win]
+    - "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}  # [not win]
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]:
```suggestion
    - {{ compiler('cxx') }}
```

File: recipes/giotto-learn/meta.yaml (hunk 4 above)
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]:
```suggestion
```

File: recipes/giotto-learn/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    - vs2017_win-64  # [win]
+    - cmake
+    - boost >=1.56.*
+    - git
+    - setuptools
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: ``boost`` and ``setuptools`` should be in host only

File: recipes/giotto-learn/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  host:
+    - python
+    - setuptools
+    - pip
+    - numpy >=1.17.0
+    - scipy
+    - scikit-learn >=0.21.3
+  run:
+    - python
+    - numpy >=1.17.0
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: Are you compiling against numpy?
if so, please see
https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=compiler%20cxx#building-against-numpy

File: recipes/giotto-learn/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    - scipy
+    - scikit-learn >=0.21.3
+
+test:
+  requires:
+    - pytest >=3.3.0
+  commands:
+    - pytest -h
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: this command should be the pytest to test this project

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: I would recommend adding the ``import`` section here
Please see
https://conda-forge.org/docs/maintainer/adding_pkgs.html?highlight=test#the-recipe-meta-yaml
https://conda-forge.org/docs/maintainer/adding_pkgs.html?highlight=test#test
https://conda-forge.org/docs/maintainer/adding_pkgs.html?highlight=test#testing-importing

File: recipes/giotto-learn/meta.yaml (hunk 2 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: We are indeed using some external packages as dependencies and we need to compile them (in particular pybind11, ripser.py and a modified version of gudhi). We add them to the project as git submodules.

File: recipes/giotto-learn/meta.yaml (hunk 6 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: done! thank you!

File: recipes/giotto-learn/meta.yaml (hunk 7 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: done, thank you!

File: recipes/giotto-learn/meta.yaml (hunk 2 above)
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: The right way is to add your dependencies in the ``host`` section and use ``cmake`` to find them

File: conda-forge.yml (hunk 1 above)
Comment (giotto-learn) [reply conda-forge.yml]: done! this was a mistake, I apologise.

File: recipes/giotto-learn/meta.yaml (hunk 8 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: import section added and working!

**Comment by giotto-learn** [comment]
Dear Conda-forge team *(in particular those with python/C++ expertise)*,
I still have this windows bug that I need help with, please. 
I have implemented all the corrections you suggested, but it seems to me that the windows build now is not compiling at all! I would like to build also for windows (as mentioned in the previous message, I can do it on my azure pipelines <https://dev.azure.com/giotto-learn/giotto-learn/_build/results?buildId=14> that look very similar to yours) but only for vs2017. My problem is that the when I put ``skip: true  # [vs<17]`` it still tries to compile with vs2015. If, on the other hand, I write ``skip: true  # [win and vs<17]``, the windows build does not compile at all. Can you please give me suggestions on how to compile only for vs2017? Thank you very much for your time!

File: recipes/giotto-learn/meta.yaml (hunk 8 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: All test are now passing (I removed the -h option)!

**Comment by giotto-learn** [comment]
Dear Conda-forge team,
Everything compiles now and all tests are passing! Eventually, can you please have a look at the windows build and check that the package is built? I am still puzzled by the effect of the selector 'vs<17' (which, I believe, is needed to compile this package).
**If everything is fine, can you please accept this pull request?**
Thank you very much for your kind attention!

File: recipes/giotto-learn/meta.yaml (hunk 2 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: these dependencies are C++ code to be bound to python by pybind11, they are not standard dependencies such as sklearn or Numpy. pybind11 requires binding files (now in the folder externals/giotto/bindings/ ) and the actual scripts. In the pybind11 examples they also clone the repositories in order to bind them: <https://github.com/pybind/cmake_example>.

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: Sorry, but ``pybind11`` is a header-only library and we have it already on ``conda-forge``

you can use on ``CMakeLists.txt`` something like
```
find_package(pybind11 REQUIRED)
```
You really do not need the ``pybind11` code to generate the bindings for python

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: Please see
https://pybind11.readthedocs.io/en/stable/compiling.html#find-package-vs-add-subdirectory

> For CMake-based projects that don’t include the pybind11 repository internally, **an external installation can be detected through find_package(pybind11)**

File: conda-forge.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (giotto-learn) [reply conda-forge.yml]: > there is probably a left-over newline.
Absolutely right: I correct it.

File: recipes/giotto-learn/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    - scikit-learn >=0.21.3
+  run:
+    - python
+    - {{ pin_compatible('numpy') }}
+    - scipy
+    - scikit-learn >=0.21.3
+
+test:
+  imports:
+    - giotto
```
[the remark this answers is not in the document]
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: I modify the import.

File: recipes/giotto-learn/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+about:
+  home: https://github.com/giotto-learn/giotto-learn
+  license: Apache-2.0
+  license_family: Apache
+  license_file: LICENSE
+  summary: 'Giotto, a topological machine learning library'
+  description: |
+    Giotto is a topological machine learning library, that is fully
+    compatible with scikit-learn. You can find more information on the
+    website <http://www.giotto.ai> : a full course, introductory videos,
```
[the remark this answers is not in the document]
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: I will then update it as soon as the website is out. Thank you!

**Comment by giotto-learn** [comment]
Thank you all for your comments and suggestion.
I have implemented most of them and now it seems that binaries are building! I am trying to make the final update concerning the pybind11 observation: remove it from submodules and use find_packages(pybind11) instead.

File: conda-forge.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rth) [reply conda-forge.yml]: It still shows in the diff -- it shouldn't...

**Comment by giotto-learn** [comment]
Dear Conda-forge team,
As you asked, I removed the pybind11 submodules: it is now a dependency!
**Can you please accept this pull request?**
Thank you very much for your kind attention.

File: conda-forge.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (giotto-learn) [reply conda-forge.yml]: I am really not sure why: is it still different?

File: recipes/giotto-learn/meta.yaml (hunk 2 above)
Comment (giotto-learn) [reply recipes/giotto-learn/meta.yaml]: Done!! Thank you for the suggestion and the links!

File: conda-forge.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (giotto-learn) [reply conda-forge.yml]: And I finally reverted the accidental modification of the `conda-forge.yml` .

**Comment by giotto-learn** [comment]
Dear Conda-forge team,
I am terribly sorry to keep on bothering you: I finally managed to revert the modification of the `conda-forge.yml` file.
**Can you please accept this pull request now?**
Thank you very much for your kind attention.

File: recipes/giotto-learn/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,62 @@
+{% set name = "giotto-learn" %}
+{% set version = "0.1a" %}
+
+package:
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  git_url: https://github.com/{{ name }}/{{ name }}.git
```
[the remark this answers is not in the document]
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos

File: recipes/giotto-learn/meta.yaml (hunk 2 above)
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: As you changed your ``CMakeLists.txt`` to find the ``pybind11`` installed, I think you do not need this line, right?
```suggestion
```

File: recipes/giotto-learn/meta.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: Could you please generate a release on github? So in that way we will be able to get this project from tarballs

File: recipes/giotto-learn/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  skip: true  # [win and vs<17]
+  script:
+    - git submodule update --init --recursive
+    - "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - cmake
+    - git
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]:
```suggestion
```

File: recipes/giotto-learn/meta.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: Related to https://github.com/giotto-learn/giotto-learn/issues/10

File: recipes/giotto-learn/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,61 @@
+{% set name = "giotto-learn" %}
+{% set version = "v0.1a.0" %}
+
+package:
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/{{ name }}/{{ name }}/tarball/{{ version }}.tar
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]:
```suggestion
  url: "https://github.com/{{ name }}/{{ name }}/archive/v{{ version }}.tar.gz"
```

File: recipes/giotto-learn/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,61 @@
+{% set name = "giotto-learn" %}
+{% set version = "v0.1a.0" %}
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]:
```suggestion
{% set version = "0.1a.0" %}
```

File: recipes/giotto-learn/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+package:
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/{{ name }}/{{ name }}/tarball/{{ version }}.tar
+  sha256: 3cd6ec8847115674700196d136b5f86c5b3b1eab64de8b040390135ff16e9349
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]:
```suggestion
  sha256: 4710ef554ab4f117fcf3953635a268526e7ea15f57f1e88655ffb425e567e646
```

File: recipes/giotto-learn/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    - cmake
+  host:
+    - python
+    - pybind11
+    - boost >=1.56.*
+    - setuptools
+    - pip
+    - numpy >=1.17.0
+    - scipy
+    - scikit-learn >=0.21.3
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: you also need ``openmp``
```suggestion
    - scikit-learn >=0.21.3
    - openmp
```

File: recipes/giotto-learn/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    - openmp
```
Comment (marcelotrevisani) [inline recipes/giotto-learn/meta.yaml]: I think you still need ``gudhi``

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=81010&view=logs&jobId=10bda42a-d420-5c28-a43f-bc4e66405e5c&taskId=bef775d0-a1a2-55a3-9721-17147029baf0&lineStart=833&lineEnd=836&colStart=34&colEnd=27

Comment (marcelotrevisani) [reply recipes/giotto-learn/meta.yaml]:
```    /tmp/pip-req-build-m7duy_h5/giotto/externals/bindings/persistent_cohomology_bindings.cpp:8:10: fatal error: gudhi/Persistent_cohomology.h: No such file or directory
     #include <gudhi/Persistent_cohomology.h>
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
```

**Pull request closed without merging**